src/main.rs

```diff
@@ -0,0 +1,272 @@
+        ));
+
+        ap.refer(&mut prometheus_push_gateway).add_option(
+            &["--prometheus-push-gateway"],
+            StoreOption,
+            "Location of the Prometheus push-gateway server to send metrics to",
+        );
+        ap.refer(&mut poll_interval).add_option(
+            &["--poll-interval"],
+            Store,
+            "Number of seconds to wait before checking mounts",
+        );
+        ap.parse_args_or_exit();
+    }
+
+    let poll_interval_duration = Duration::from_secs(poll_interval);
+
+    println!(
+        "mount_status_monitor checking mounts every {} seconds",
+        poll_interval_duration.as_secs()
+    );
+
+    let syslog = syslog::unix(Facility::LOG_DAEMON).unwrap_or_else(|err| {
+        eprintln!("Unable to connect to syslog: {}", err);
+        std::process::exit(1);
+    });
+
+    let mut mount_statuses = HashMap::<String, MountStatus>::new();
+
+    loop {
+        check_mounts(&mut mount_statuses, &syslog);
+
+        // We calculate these values each time because a filesystem may have been
+        // mounted or unmounted since the last check:
+        let total_mounts = mount_statuses.len();
+        let dead_mounts = mount_statuses
+            .iter()
+            .map(|(_, status)| !status.alive)
+            .filter(|&i| i)
+            .count();
+
+        // TODO: consider making this debug or sending it to stdout?
+        syslog
+            .info(format!(
+                "Checked {} mounts; {} are dead",
+                total_mounts,
+                dead_mounts
+            ))
+            .unwrap_or_else(handle_syslog_error);
+
+        if prometheus_push_gateway.is_some() {
+            let prometheus_instance = hostname::get_hostname().unwrap();
+
+            // The Prometheus metrics are defined as floats so we need to convert;
+            // for monitoring the precision loss in general is fine and it's
+            // exceedingly unlikely to be relevant when counting the number of
+            // mountpoints:
+            TOTAL_MOUNTS.set(total_mounts as f64);
+            DEAD_MOUNTS.set(dead_mounts as f64);
+
+            if let Err(e) = prometheus::push_metrics(
+                "mount_status_monitor",
+                labels!{"instance".to_owned() => prometheus_instance.to_owned(), },
+                prometheus_push_gateway.as_ref().unwrap(),
+                prometheus::gather(),
+            )
+            {
+                eprintln!("Unable to send pushgateway metrics: {}", e);
+            }
+        }
+
+        // Wait before checking again:
+        thread::sleep(poll_interval_duration);
+    }
+}
+
+fn check_mounts(mount_statuses: &mut HashMap<String, MountStatus>, logger: &syslog::Logger) {
+    let mount_points = get_mounts::get_mount_points();
+
+    // Remove any mount status entries which are no longer in the current list of mountpoints:
+    mount_statuses.retain(|ref k, _| {
+        mount_points.iter().position(|i| *i == **k).is_some()
```

Comment:
I guess, If you remove the `ref` on `ref k` above, then you can get rid of one asterisk before `**k`

Reply:
Could be worthwhile to use a BTreeSet/HashSet, or a sorted list + Binary Search for the set of mount points.

Reply:
@Dr-Emann I wasn't sure how much I should care about that given that the list of mounts is usually well under a hundred items but it might warrant further thought.

Reply:
Additionally, it can also remove any duplicates, if you mount something over another mount (`mount -t tmpfs tmpfs /mypath; mount -t tmpfs tmpfs /mypath`, currently we'll do 2 checks for /mypath)